pkg/depend/write.go

```diff
@@ -1,12 +1,75 @@
 package depend
 
-import "github.com/git-depend/git-depend/pkg/git"
+import (
+	"encoding/json"
+	"errors"
+	"time"
 
-// Write a request to a note.
-func (req *Request) Write(cache *git.Cache, ref string) error {
-	json, err := req.GetJson()
+	"github.com/git-depend/git-depend/pkg/git"
+	"github.com/git-depend/git-depend/pkg/utils"
+)
+
+type Status int
+
+const (
+	Locked Status = iota
+	Unlocked
+)
+
+func (s Status) String() string {
+	return [...]string{"Locked", "Unlocked"}[s]
+}
+
+var ref_lock_name string = "git-depend-lock"
+
+// Lock allows us to safely write to a note.
+type Lock struct {
+	ID        string    `json:"Id"`
+	Timestamp time.Time `json:"Timestamp"`
+	Status    string    `json:"Status"`
+}
+
+// WriteLock will lock an individual repository.
+func (lock *Lock) WriteLock(cache *git.Cache, node *Node) error {
+	data, err := json.Marshal(lock)
 	if err != nil {
 		return err
 	}
-	return cache.ForceAddNotes(req.Url, ref, string(json))
+
+	if err = cache.AddNotes(node.URL, ref_lock_name, string(data)); err != nil {
+		return err
+	}
+
+	return cache.PushNotes(node.URL, ref_lock_name)
+}
+
+// WriteLocks for a request and the children.
+func (requests *Requests) WriteLocks(cache *git.Cache) error {
+	visited := utils.NewSet()
+	lock := &Lock{
+		ID:        "foo",
+		Timestamp: time.Now(),
+		Status:    Locked.String(),
+	}
+	for k := range requests.Table {
+		node, ok := requests.NodesRoot.GetNode(k)
+		if !ok {
+			return errors.New("Key does not exist: " + k)
+		}
+		if !visited.Exists(k) {
+			if err := lock.WriteLock(cache, node); err != nil {
```

Comment:
Can we do atomic pushes of git notes (see [here](https://git-scm.com/docs/git-push#Documentation/git-push.txt---no-atomic)?) If so, this might make this a bit cleaner as presumably, we will want to lock all repos atomically if the merge transaction is also going to proceed atomically?

Reply:
We only write one note per repo so it probably isn't a problem and I don't want to depend on the server implementing this.